lib/maths/CBayesianOptimisation.cc

```diff
@@ -189,6 +205,176 @@ CBayesianOptimisation::maximumExpectedImprovement() {
     return {std::move(xmax), expectedImprovement};
 }
 
+CBayesianOptimisation::TVector CBayesianOptimisation::kinvf() const {
+    TVector Kinvf;
+    TMatrix K{this->kernel(m_KernelParameters, this->meanErrorVariance())};
+    Kinvf = K.ldlt().solve(this->function());
+    return Kinvf;
+}
+
+CBayesianOptimisation::TVector CBayesianOptimisation::transformTo01(const TVector& x) const {
+    return (x - m_MinBoundary).cwiseQuotient(m_MaxBoundary - m_MinBoundary);
+}
+
+CBayesianOptimisation::TVector CBayesianOptimisation::scaledKernelParameters() const {
+    TVector scaledKernelParameters{m_KernelParameters.tail(m_MinBoundary.size())};
+    return scaledKernelParameters.cwiseProduct(m_MaxBoundary - m_MinBoundary);
+}
+
+double CBayesianOptimisation::evaluate(const TVector& input) const {
+    return this->evaluate(this->kinvf(), input);
+}
+
+double CBayesianOptimisation::evaluate(const TVector& Kinvf, const TVector& input) const {
+    TVector Kxn;
+    std::tie(Kxn, std::ignore) = this->kernelCovariates(m_KernelParameters, input,
+                                                        this->meanErrorVariance());
+    return Kxn.transpose() * Kinvf;
+}
+
+double CBayesianOptimisation::evaluate1D(const TVector& Kinvf, double input, int dimension) const {
+    TVector scaledKernelParameters{this->scaledKernelParameters()};
+    auto prodXt = [&scaledKernelParameters, this](const TVector& x, int t) {
+        double prod{1.0};
+        for (int d = 0; d < m_MinBoundary.size(); ++d) {
+            if (d != t) {
+                prod *= integrate1dKernel(scaledKernelParameters(d), x(d));
+            }
+        }
+        return prod;
+    };
+    double f0{this->anovaConstantFactor(Kinvf)};
```

Comment:
As before, since we're differencing numbers proportional to theta_0^4 in some places I'm a little worried about cancellation errors. Using here as an example of what I have in mind (same approach applies elsewhere), I would propose introducing a scale. For example, the following changes to:
```
    double theta02{CTools::pow2(m_KernelParameters(0))};
    double scale{std::max(1.0, theta02)};
    double f0{this->anovaConstantFactor(Kinvf) / scale};
    double sum{0.0};
    input = (input - m_MinBoundary(dimension)) /
            (m_MaxBoundary(dimension) - m_MinBoundary(dimension));
    for (std::size_t i = 0; i < m_FunctionMeanValues.size(); ++i) {
        TVector x{this->transformTo01(m_FunctionMeanValues[i].first)};
        sum += Kinvf(static_cast<int>(i)) *
               std::exp(-(CTools::pow2(scaledKernelParameters[dimension]) +
                          MINIMUM_KERNEL_COORDINATE_DISTANCE_SCALE) *
                        CTools::pow2(input - x(dimension))) *
               prodXt(x, dimension);
    }
    return scale * (theta02 / scale * sum - f0);
```
I feel this doesn't complicate matters too much and deals with the edge case. Of course it only really matters if `|f0| >> |CTools::pow2(m_KernelParameters(0)) * sum - f0|`. Also these are approximations of the quantities we care about anyway and given we're using double precision we probably get away with it. I'll leave it to you to decide whether to roll out these sorts of change.

Reply:
Better safe than sorry. I added your suggestion to the relevant functions.